taint-mode: Specify sources/sanitizers/sinks using pattern formulas

IagoAbal opened the pull request:
We are no longer restricted to single instruction/expression patterns.

Goodbye fake taint?

Helps #3199

test plan:
make test # tests included



PR checklist:
- [x] changelog is up to date

aryx reviewed (approved):
Fantastic!

aryx commented on semgrep-core/tests/tainting_rules/go/zip-traversal.yaml:
```diff
@@ -0,0 +1,26 @@
+            ...
+    pattern-sinks:
+      - patterns:
+        - pattern: $PATH
```
Will this be optimized or the recent revert of emma will prevent the optimization here?

IagoAbal replied:
Unfortunately this won't be optimized now, we need to put the optimization back or find some other alternative if we want this to be efficient.

aryx commented on semgrep-core/src/engine/Tainting_generic.ml:
```diff
@@ -47,61 +46,40 @@ module DataflowY = Dataflow.Make (struct
-
-(*e: function [[Tainting_generic.match_pat_instr]] *)
+let any_in_ranges any ranges =
+  let tok1, tok2 = Visitor_AST.range_of_any any in
```
Maybe add a comment that says "This is potentially slow. We may need to store
range position in the AST at some point".

IagoAbal replied:
Added!

IagoAbal committed:
taint-mode: Specify sources/sanitizers/sinks using pattern formulas

We are no longer restricted to single instruction/expression patterns.

Goodbye fake taint?

Helps #3199

test plan:
make test # tests included
